src/Http/WebUtilities/src/FileBufferingSectionReadStream.cs

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// A Stream that wraps another stream and enables rewinding by buffering the content as it is read.
+    /// The content is buffered in memory up to a certain size and then spooled to a temp file on disk.
+    /// The temp file will be deleted on Dispose.
+    /// </summary>
+    public class FileBufferingPipeReaderStream : Stream
```

Comment:
Do we really need a whole new type for this?

Reply:
I suspect not, it's a matter of layering things correctly. This could be replaced by the normal Pipe->Stream adapter and the existing buffer stream.

Reply:
The reason for having a new type is that the other FileBufferingStream is using a Stream internally. I wanted to keep the code from previous functionality to still be there - but I guess that FileBufferingStream could be removed.

I wanted to not use a Pipe-> Stream adapter because I guess there should be some performance benefits to implementing ReadAsString() or DrainAsync without allocating a byte array.